libsolutil/Algorithms.h

```diff
@@ -106,11 +106,11 @@ struct BreadthFirstSearch
 			V v = std::move(verticesToTraverse.front());
 			verticesToTraverse.pop_front();
 
-			visited.insert(v);
+			if (!visited.insert(v).second)
+				continue;
 
 			_forEachChild(v, [this](V _vertex) {
-				if (!visited.count(_vertex))
```

Comment:
If we also check here, we can reduce the size of the list. Is that worth it?

Reply:
Also, is any client using the `visited` member? If not, I would suggest to change it to 'queued' instead.

Reply:
Almost all clients *only* use this for getting the ``visited`` member :-).

Reply:
I'd argue checking the list again here is ``O(log n)`` whereas appending and removing from the list is ``O(1)``, so adding here and removing later, if already visited then, is better, since it saves an ``O(log n)`` search in all cases in which the child was *not* already visited, whereas having a longer list is negligible...
However, one could also argue that appending to the list has *high* constant cost, since it actually even allocates, etc... maybe a ``std::deque`` would be better than a ``std::list`` in that regard...

Anyways: seems fine either way to me.

Reply:
ok!